WordPress/gutenberg
# Use block naming for marking blocks as overridable in patterns

## What?
Attempts to switch over to using the block metadata name instead of a nano id as the connecting glue for binding pattern overrides.

## How?
- Removes the 'allow instance overrides' checkbox
- Removes the 'blockName' property from the pattern content data
- Replaces the `PartialSyncingControls` component with a hook that monitors whether a metadata name for a block has been entered and appropriately adds/removes the bindings for blocks
- Removes the `values` sub-property in the pattern block `content` attribute
- Updates the PHP block binding callback to use the metadata name instead of id for bindings (requires a backport)
- Adds some migrations and other general backwards compatibility

## Current issues / questions
- What do we do when there are existing names for blocks, should bindings be added automatically? At the moment the user has to remove and re-add the name.
- Does using only the names give the user enough control over whether they want a binding?
- It's very easy to break a binding by renaming a block - some guiderails/warnings may have to be added to prevent this

## Testing Instructions
1. Create a pattern and name it something like 'New pattern'
2. Name for any blocks that should have overrides
3. Save the pattern
4. Insert the pattern into a post and edit the overriden blocks
5. Preview the post and check that the frontend works

### Backwards compatibility testing
1. Repeat the steps above, but with `trunk` checked out and name the pattern something like 'Old pattern' 😄 
2. Save the post that you inserted the pattern into
3. Check out this branch, build
4. Reload the post that had the pattern inserted, it should show as before
5. Preview the post, it should render the same on the frontend

## Screenshots or screencast
